tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -850,5 +850,589 @@ int main(int argc, char **argv)
+                 * then server will attempt to first negotiate the bottom two ciphersuites and then try to negotiate its third most-preferred
+                 * ciphersuite.
+                 */
+                static struct s2n_cipher_suite *cipher_suites_with_tls12_chacha20[] = {
+                    &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256,       
+                    &s2n_ecdhe_ecdsa_with_chacha20_poly1305_sha256, 
+                    &s2n_ecdhe_rsa_with_aes_256_gcm_sha384,         
+                    &s2n_ecdhe_rsa_with_chacha20_poly1305_sha256,   
+                    &s2n_dhe_rsa_with_chacha20_poly1305_sha256,     
+                };
+
+                cipher_preferences.count = s2n_array_len(cipher_suites_with_tls12_chacha20);
+                cipher_preferences.suites = cipher_suites_with_tls12_chacha20;
+
+                /*
+                 * 1. most preferred ciphersuite uses chacha20 alg. Client has signalled chacha20 boosting. (also supported by the server)
+                 * 2. second preferred is ecdsa based chacha20 alg. While this is the server's most preferred chacha20 ciphersuite,
+                 *    it is unable to negotiate it since the server does not support ecdsa auth.
+                 * 3. third preferred is dhe_rsa_chacha20 ciphersuite, the server's least preferred ciphersuite
+                 * 4. fourth preferred is an ecdsa ciphersuite; unnegotiable by server
+                 * 5. least preferred is an rsa_aes_256 ciphersuite, negotiable by server
+                 */
+                uint8_t dhe_chacha20_boosted_wire[] = {
+                    TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+                    TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
+                    TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+                    TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,
+                    TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+                };
+                uint8_t count = sizeof(dhe_chacha20_boosted_wire) / S2N_TLS_CIPHER_SUITE_LEN;
+
+                cipher_preferences.allow_chacha20_boosting = true;
+                EXPECT_TRUE(cipher_preferences.allow_chacha20_boosting);
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, dhe_chacha20_boosted_wire, count));
+                /* Server prefers ecdsa chacha20 but the connection only supports rsa. Servers moves on to the next chacha20 ciphersuite */
+                EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_ecdhe_rsa_with_chacha20_poly1305_sha256);
+
+                cipher_preferences.allow_chacha20_boosting = false;
+                EXPECT_FALSE(cipher_preferences.allow_chacha20_boosting);
+
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, dhe_chacha20_boosted_wire, count));
+                EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_ecdhe_rsa_with_aes_256_gcm_sha384);
```

Comment:
Can we make it clearer what tests / assertions we're interested in here?
You also don't need to expect the value of a bool you just set. We're paranoid, but you can trust a basic assignment :)
```suggestion
                /* Verify chacha20 RSA cipher suite chosen with chacha20 boosting enabled */
                cipher_preferences.allow_chacha20_boosting = true;
                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, dhe_chacha20_boosted_wire, count));
                EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_ecdhe_rsa_with_chacha20_poly1305_sha256);

                /* Sanity check: non-chacha20 RSA cipher suite chosen without chacha20 boosting enabled */
                cipher_preferences.allow_chacha20_boosting = false;
                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, dhe_chacha20_boosted_wire, count));
                EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_ecdhe_rsa_with_aes_256_gcm_sha384);
```

Reply:
I'll add the comments. About the asserts; I tried using them as a way to get around cppcheck's `redundantAssignment` complaint. What do you recommend instead?

Reply:
That would depend on why cppcheck is complaining about redundantAssignment. You might have to experiment.

Maybe cppcheck would be happy if you just didn't explicitly declare allow_chacha20_boosting:
```
cipher_preferences = (struct s2n_cipher_preferences) {
    .count = s2n_array_len(test_cipher_suite_list),
    .suites = test_cipher_suite_list,
};
```

Or just only set the fields you care about:
```
cipher_preferences.suites = test_cipher_suite_list;
cipher_preferences.count = s2n_array_len(test_cipher_suite_list);
```
Or, failing all of that, do the EXPECT once with a comment explaining why it's necessary:
```
cipher_preferences = (struct s2n_cipher_preferences) {
     .count = s2n_array_len(test_cipher_suite_list),
    .suites = test_cipher_suite_list,
     .allow_chacha20_boosting = true
};
/* To avoid a cppcheck redundantAssignment */
EXPECT_TRUE(cipher_preferences.allow_chacha20_boosting);
```
But this seems avoidable.

Reply:
I pulled down your code to see the actual cppcheck errors. It looks like it's complaining about the second direct assignment in each test, the one you do for your sanity checks.

I'd just stick a `/* cppcheck-suppress redundantAssignment */` on those specific lines. You can search the codebase for "cppcheck-suppress" to see examples of in-line suppression.

Reply:
Thanks, let me try